quinn-proto/src/connection/streams/recv.rs

```diff
@@ -304,24 +304,29 @@ impl<'a> Chunks<'a> {
             !drop || matches!(state, ChunksState::Finalized),
             "finalize must be called before drop"
         );
-        match state {
+        if matches!(state, ChunksState::Finalized) {
+            // Noop on repeated calls
+            return ShouldTransmit(false);
+        }
+
+        let mut should_transmit = match state {
             ChunksState::Readable(mut rs) => {
                 let (_, max_stream_data) = rs.max_stream_data(self.streams.stream_receive_window);
-                let max_data = self.streams.add_read_credits(self.read);
                 self.pending
-                    .post_read(self.id, max_data, max_stream_data, false);
+                    .post_read(self.id, ShouldTransmit(false), max_stream_data, false);
                 self.streams.recv.insert(self.id, rs);
-                ShouldTransmit(max_stream_data.0 | max_data.0)
-            }
-            ChunksState::Finished => {
-                // MAX_DATA may need to be issued, but MAX_STREAM_DATA is pointless
-                let max_data = self.streams.add_read_credits(self.read);
-                self.pending.max_data |= max_data.0;
-                max_data
+                max_stream_data.0
             }
-            ChunksState::Reset(_) => ShouldTransmit(true),
-            ChunksState::Finalized => ShouldTransmit(false),
-        }
+            ChunksState::Finished => false,
+            ChunksState::Reset(_) => true,
+            ChunksState::Finalized => false,
+        };
+
+        // Issue connection-level flow control credit for any data we read regardless of state
+        let max_data = self.streams.add_read_credits(self.read);
+        self.pending.max_data |= max_data.0;
+        should_transmit |= max_data.0;
+        ShouldTransmit(should_transmit)
```

Comment:
There actually isn't a semantic change. Consider each state:
- Readable: Credit issued before and after the change.
- Finished: Credit issued before and after the change.
- Reset: Before the change, credit never issued because `add_read_credits` isn't called. After the change, credit never issued because `self.read` is necessarily 0, because we drop all assembler buffers immediately upon receiving a reset, therefore no data can have been read. This is pretty subtle, so I added a `debug_assert` to clarify and extended a test to validate.
- Finalized: Credit never issued before or after the change because `add_read_credits` isn't called.

Reply:
Ah, very cool. Thanks!